File: EduTech/EduTech/Shared/MathText.swift
import SwiftUI
import WebKit

/// Renders text with LaTeX math (\(...\) and \[...\]) via bundled KaTeX (no CDN).
/// Falls back to plain Text if no math markers detected.
struct MathText: View {
    let text: String
    var fontSize: CGFloat = 17

    @State private var webHeight: CGFloat = 60

    private var hasMath: Bool {
        text.contains("\\(") || text.contains("\\[")
    }

    var body: some View {
        if hasMath {
            _MathWebView(text: text, fontSize: fontSize, height: $webHeight)
                .frame(height: webHeight)
                .frame(maxWidth: .infinity, alignment: .leading)
        } else {
            Text(text)
                .font(.system(size: fontSize, design: .serif))
                .fixedSize(horizontal: false, vertical: true)
                .frame(maxWidth: .infinity, alignment: .leading)
        }
    }
}

private struct _MathWebView: UIViewRepresentable {
    let text: String
    let fontSize: CGFloat
    @Binding var height: CGFloat

    func makeCoordinator() -> Coordinator { Coordinator(height: $height) }

    // Points to the bundled KaTeX folder so scripts/fonts load from disk, no network needed
    private static let katexBase: URL = {
        Bundle.main.url(forResource: "katex.min", withExtension: "js")?
            .deletingLastPathComponent() ?? Bundle.main.bundleURL
    }()

    func makeUIView(context: Context) -> WKWebView {
        let wv = WKWebView()
        wv.navigationDelegate = context.coordinator
        wv.isOpaque = false
        wv.backgroundColor = .clear
        wv.scrollView.isScrollEnabled = false
        wv.scrollView.bounces = false
        context.coordinator.lastText = text
        wv.loadHTMLString(html(text), baseURL: Self.katexBase)
        return wv
    }

    func updateUIView(_ wv: WKWebView, context: Context) {
        guard context.coordinator.lastText != text else { return }
        context.coordinator.lastText = text
        wv.loadHTMLString(html(text), baseURL: Self.katexBase)
    }

    static func dismantleUIView(_ uiView: WKWebView, coordinator: Coordinator) {
        uiView.stopLoading()
        uiView.navigationDelegate = nil
    }

    private func html(_ raw: String) -> String {
        let body = raw
            .replacingOccurrences(of: "&", with: "&amp;")
            .replacingOccurrences(of: "<", with: "&lt;")
            .replacingOccurrences(of: ">", with: "&gt;")
            .replacingOccurrences(of: "\n", with: "<br>")
        return """
        <!DOCTYPE html><html><head>
        <meta name="viewport" content="width=device-width,initial-scale=1,maximum-scale=1,user-scalable=no">
        <link rel="stylesheet" href="katex.min.css">
        <script defer src="katex.min.js"></script>
        <script defer src="auto-render.min.js"
            onload="renderMathInElement(document.body,{delimiters:[{left:'\\\\(',right:'\\\\)',display:false},{left:'\\\\[',right:'\\\\]',display:true}],throwOnError:false})"></script>
        <style>
        *{box-sizing:border-box}
        body{margin:0;padding:2px 0;font-family:-apple-system,Georgia,serif;font-size:\(Int(fontSize))px;line-height:1.5;word-wrap:break-word;background:transparent}
        @media(prefers-color-scheme:dark){body{color:#f5f5f5}}
        @media(prefers-color-scheme:light){body{color:#111}}
        .katex{font-size:1.05em}
        </style>
        </head><body>\(body)</body></html>
        """
    }

    final class Coordinator: NSObject, WKNavigationDelegate {
        @Binding var height: CGFloat
        var lastText: String = ""

        init(height: Binding<CGFloat>) { _height = height }

        func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
            // Measure twice: initial layout + after KaTeX scripts finish rendering
            measure(webView, after: 0.1)
            measure(webView, after: 0.8)
        }

        private func measure(_ wv: WKWebView, after delay: TimeInterval) {
            DispatchQueue.main.asyncAfter(deadline: .now() + delay) { [weak self, weak wv] in
                wv?.evaluateJavaScript("document.body.scrollHeight") { result, _ in
                    let h: CGFloat
                    if let d = result as? Double { h = CGFloat(d) }
                    else if let i = result as? Int { h = CGFloat(i) }
                    else { return }
                    if h > 10 { self?.height = h + 4 }
                }
            }
        }
    }
}

File: app/src/main/java/dev/squad52/android_edutech/ui/components/MathText.kt
package dev.squad52.android_edutech.ui.components

import android.graphics.Color as AndroidColor
import android.webkit.WebView
import android.webkit.WebViewClient
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.material3.Text
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.toArgb
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.compose.ui.viewinterop.AndroidView
import dev.squad52.android_edutech.ui.theme.LocalAppColors
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

@Composable
fun MathText(
    text: String,
    modifier: Modifier = Modifier,
    fontSize: Int = 16,
    color: Color? = null
) {
    val colors = LocalAppColors.current
    val fgColor = color ?: colors.foreground
    val hasMath = text.contains("\\(") || text.contains("\\[")

    if (!hasMath) {
        Text(
            text = text,
            modifier = modifier,
            fontSize = fontSize.sp,
            color = fgColor,
            lineHeight = (fontSize * 1.5f).sp
        )
        return
    }

    val colorHex = "#%06X".format(fgColor.toArgb() and 0xFFFFFF)

    key(text, colorHex) {
        var webHeight by remember { mutableStateOf(80.dp) }
        val scope = rememberCoroutineScope()

        AndroidView(
            factory = { ctx ->
                WebView(ctx).apply {
                    settings.javaScriptEnabled = true
                    @Suppress("DEPRECATION")
                    settings.allowUniversalAccessFromFileURLs = true
                    settings.allowFileAccess = true
                    setBackgroundColor(AndroidColor.TRANSPARENT)
                    isScrollContainer = false

                    webViewClient = object : WebViewClient() {
                        override fun onPageFinished(view: WebView, url: String) {
                            scope.launch {
                                delay(150)
                                view.evaluateJavascript("document.body.scrollHeight") { h ->
                                    h?.trim('"')?.toDoubleOrNull()?.let { px ->
                                        webHeight = (px + 8).dp
                                    }
                                }
                                delay(750)
                                view.evaluateJavascript("document.body.scrollHeight") { h ->
                                    h?.trim('"')?.toDoubleOrNull()?.let { px ->
                                        webHeight = (px + 8).dp
                                    }
                                }
                            }
                        }
                    }

                    loadDataWithBaseURL(
                        "file:///android_asset/katex/",
                        buildMathHtml(text, colorHex, fontSize),
                        "text/html",
                        "UTF-8",
                        null
                    )
                }
            },
            modifier = modifier.fillMaxWidth().height(webHeight)
        )
    }
}

private fun buildMathHtml(text: String, colorHex: String, fontSize: Int): String {
    val escaped = text
        .replace("&", "&amp;")
        .replace("<", "&lt;")
        .replace(">", "&gt;")
    return """<!DOCTYPE html>
<html>
<head>
<meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0">
<link rel="stylesheet" href="katex.min.css">
<script defer src="katex.min.js"></script>
<script defer src="auto-render.min.js"
    onload="renderMathInElement(document.body,{delimiters:[{left:'\\(',right:'\\)',display:false},{left:'\\[',right:'\\]',display:true}],throwOnError:false});">
</script>
<style>
* { margin: 0; padding: 0; box-sizing: border-box; }
html, body { width: 100%; max-width: 100%; overflow-x: hidden; }
body { font-family: sans-serif; font-size: ${fontSize}px; color: $colorHex; line-height: 1.5; background: transparent; padding: 2px 0; word-break: break-word; }
.katex { font-size: 1em; }
.katex-display { overflow-x: auto; overflow-y: hidden; }
img, svg { max-width: 100%; height: auto; }
</style>
</head>
<body>$escaped</body>
</html>"""
}
